src/batchmap.rs

```diff
@@ -0,0 +1,627 @@
+    /// Default value is `/var/run/numaflow/batchmap.sock`
+    pub fn with_socket_file(mut self, file: impl Into<PathBuf>) -> Self {
+        self.sock_addr = file.into();
+        self
+    }
+
+    /// Get the unix domain socket file path where gRPC server listens for incoming connections. Default value is `/var/run/numaflow/batchmap.sock`
+    pub fn socket_file(&self) -> &std::path::Path {
+        self.sock_addr.as_path()
+    }
+
+    /// Set the maximum size of an encoded and decoded gRPC message. The value of `message_size` is in bytes. Default value is 64MB.
+    pub fn with_max_message_size(mut self, message_size: usize) -> Self {
+        self.max_message_size = message_size;
+        self
+    }
+
+    /// Get the maximum size of an encoded and decoded gRPC message in bytes. Default value is 64MB.
+    pub fn max_message_size(&self) -> usize {
+        self.max_message_size
+    }
+
+    /// Change the file in which numaflow server information is stored on start up to the new value. Default value is `/var/run/numaflow/batchmapper-server-info`
+    pub fn with_server_info_file(mut self, file: impl Into<PathBuf>) -> Self {
+        self.server_info_file = file.into();
+        self
+    }
+
+    /// Get the path to the file where numaflow server info is stored. Default value is `/var/run/numaflow/mapper-server-info`
+    pub fn server_info_file(&self) -> &std::path::Path {
+        self.server_info_file.as_path()
+    }
+
+    /// Starts the gRPC server. When message is received on the `shutdown` channel, graceful shutdown of the gRPC server will be initiated.
+    pub async fn start_with_shutdown(
+        &mut self,
+        shutdown_rx: oneshot::Receiver<()>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+    where
+        T: BatchMapper + Send + Sync + 'static,
+    {
+        let mut info = shared::default_info_file();
+        // update the info json metadata field, and add the map mode
+        info["metadata"][shared::MAP_MODE_KEY] =
+            serde_json::Value::String(shared::BATCH_MAP.to_string());
+        let listener =
+            shared::create_listener_stream(&self.sock_addr, &self.server_info_file, info)?;
+        let handler = self.svc.take().unwrap();
+
+        // Create a channel to send shutdown signal to the server to do graceful shutdown in case of non retryable errors.
+        let (internal_shutdown_tx, internal_shutdown_rx) = mpsc::channel(1);
+        let map_svc = crate::batchmap::BatchMapService {
+            handler,
+            _shutdown_tx: internal_shutdown_tx,
+        };
+
+        let map_svc = proto::batch_map_server::BatchMapServer::new(map_svc)
+            .max_encoding_message_size(self.max_message_size)
+            .max_decoding_message_size(self.max_message_size);
+
+        let shutdown = shutdown_signal(
+            internal_shutdown_rx,
+            Some(shutdown_rx),
+            CancellationToken::new(),
```

Comment:
The `CancellationToken` is intended to cancel the tasks spawned by the `BatchMapService` when the Grpc server gets a shutdown signal. It seems only `Reducer` implementation uses it as of now. We might want to start implementing it for all.

Reply:
that is right, we need to do for rest.